### PR TITLE
xds: add panic recovery in xdsclient resource unmarshalling.

### PR DESCRIPTION
**Pull request by Pranjali-2501**
This PR introduces panic recovery mechanisms to the xdsclient resource unmarshaling logic. 

Currently, a panic during the parsing of a malformed xDS update could crash the entire application. To improve resilience in production environments, this change wraps the resource parsing logic in a `defer recover()` block.

The recovery mechanism is controlled by the `GRPC_EXPERIMENTAL_XDS_PANIC_RECOVERY` environment variable, which defaults to `true`.
- When enabled, panics during unmarshaling are caught and returned as errors, preventing the application from crashing.
- When disabled, panics are allowed to propagate. This is required for fuzz testing, where crashing on unexpected input is the desired behavior to identify bugs: [go/grpc-go-xdsclient-fuzzing](http://goto.google.com/grpc-go-xdsclient-fuzzing)

Part of : #8749 

RELEASE NOTES: N/A

### CODE REVIEW

File: internal/xds/xdsclient/xdsresource/unmarshal_cds.go [hunk 1]
```diff
@@ -51,8 +51,15 @@ var ValidateClusterAndConstructClusterUpdateForTesting = validateClusterAndConst
 // to this value by the management server.
 const transportSocketName = "envoy.transport_sockets.tls"
 
-func unmarshalClusterResource(r *anypb.Any, serverCfg *bootstrap.ServerConfig) (string, ClusterUpdate, error) {
-	r, err := UnwrapResource(r)
+func unmarshalClusterResource(r *anypb.Any, serverCfg *bootstrap.ServerConfig) (name string, update ClusterUpdate, err error) {
```
Comment (eshitachandwani) [inline internal/xds/xdsclient/xdsresource/unmarshal_cds.go]: Can you add to the description example of where the panics can occur? From the unmarshal functions, it looks like all the edge cases are handled by returning appropriate errors.

Comment (Pranjali-2501) [reply internal/xds/xdsclient/xdsresource/unmarshal_cds.go]: We don't have any specific test case that triggers a panic right now, as the current unmarshal functions do handle known edge cases by returning errors.

However, we have decided to add this recover purely as a defensive measure for production stability.

**Comment by eshitachandwani** [comment]
Re-assigning to @Pranjali-2501 to addresses Easwar's comments from the CL.

File: internal/xds/clients/xdsclient/channel.go [hunk 2]
```diff
@@ -253,7 +254,16 @@ func decodeResponse(opts *DecodeOptions, rType *ResourceType, resp response) (ma
 	perResourceErrors := make(map[string]error) // Tracks resource validation errors, where we have a resource name.
 	ret := make(map[string]dataAndErrTuple)     // Return result, a map from resource name to either resource data or error.
 	for _, r := range resp.resources {
-		result, err := rType.Decoder.Decode(NewAnyProto(r), *opts)
+		result, err := func() (res *DecodeResult, err error) {
+			defer func() {
+				if envconfig.XDSRecoverPanicInResourceParsing {
+					if p := recover(); p != nil {
+						err = fmt.Errorf("panic during %v resource decoding: %v", rType.TypeName, p)
```
[the remark this answers is not in the document]
Comment (Pranjali-2501) [reply internal/xds/clients/xdsclient/channel.go]: Done.

File: internal/envconfig/envconfig.go [hunk 3]
```diff
@@ -96,6 +96,17 @@ var (
 	// feature can be disabled by setting the environment variable
 	// GRPC_EXPERIMENTAL_PF_WEIGHTED_SHUFFLING to "false".
 	PickFirstWeightedShuffling = boolFromEnv("GRPC_EXPERIMENTAL_PF_WEIGHTED_SHUFFLING", true)
+
+	// XDSRecoverPanicInResourceParsing indicates whether the xDS resolver should
+	// recover from panics while parsing xDS resources.
+	//
+	// The recover logic in xdsclient resource parsing is protected by an
+	// environment variable so that it can be disabled during fuzz testing to
```
Comment (eshitachandwani) [inline internal/envconfig/envconfig.go]: We do not need the middle line , it feels repetitive. You can mention the fuzz testing in the line below and say something like : 
```
This feature can be disabled (e.g., for fuzz testing) by setting the `GRPC_EXPERIMENTAL_XDS_PANIC_RECOVERY` environment variable to false."
```

File: internal/xds/clients/xdsclient/channel_test.go [hunk 4]
```diff
@@ -772,3 +774,61 @@ func (ta *testEventHandler) waitForResourceDoesNotExist(ctx context.Context) (Re
...
 }
+
+type panicDecoder struct{}
+
+func (panicDecoder) Decode(*AnyProto, DecodeOptions) (*DecodeResult, error) {
+	panic("simulate panic")
+}
+
+// TestDecodeResponse_PanicRecovery tests the panic recovery mechanism in
+// decodeResponse. It verifies that when XDSRecoverPanicInResourceParsing env variable is
```
Comment (eshitachandwani) [inline internal/xds/clients/xdsclient/channel_test.go]: Can you please wrap this comment to 80 cols.

File: internal/xds/clients/xdsclient/channel_test.go [hunk 5]
```diff
@@ -772,3 +774,61 @@ func (ta *testEventHandler) waitForResourceDoesNotExist(ctx context.Context) (Re
...
+	tests := []struct {
+		name          string
+		enableRecover bool
+		wantPanic     bool
+		wantErr       string
+	}{
+		{
+			name:          "Enable_XDSRecoverPanic",
+			enableRecover: true,
+			wantPanic:     false,
```
Comment (eshitachandwani) [inline internal/xds/clients/xdsclient/channel_test.go]: We can skip the false fields since default initialisation is false.

File: internal/xds/clients/xdsclient/channel_test.go [hunk 6]
```diff
@@ -772,3 +774,61 @@ func (ta *testEventHandler) waitForResourceDoesNotExist(ctx context.Context) (Re
...
+				resources: []*anypb.Any{{Value: []byte("test")}},
+			}
+
+			if tt.wantPanic {
+				defer func() {
+					if r := recover(); r == nil {
+						t.Errorf("The code did not panic but was expected to")
+					}
+				}()
+				decodeResponse(opts, rType, resp)
```
Comment (eshitachandwani) [inline internal/xds/clients/xdsclient/channel_test.go]: Can we return from here so that we do not need to write the rest of the code in an else block.

File: internal/xds/clients/xdsclient/channel_test.go [hunk 7]
```diff
@@ -772,3 +774,61 @@ func (ta *testEventHandler) waitForResourceDoesNotExist(ctx context.Context) (Re
...
+			} else {
+				if _, _, err := decodeResponse(opts, rType, resp); err == nil || !strings.Contains(err.Error(), tt.wantErr) {
+					t.Errorf("decodeResponse() failed with err: %v, want %q", err, tt.wantErr)
```
Comment (eshitachandwani) [inline internal/xds/clients/xdsclient/channel_test.go]: Since this is the last check , we can change it to use `t.Fatalf`

File: internal/xds/clients/xdsclient/channel_test.go [hunk 8]
```diff
@@ -772,3 +774,61 @@ func (ta *testEventHandler) waitForResourceDoesNotExist(ctx context.Context) (Re
...
+				TypeName: "resourceType",
+				Decoder:  panicDecoder{},
+			}
+			resp := response{
+				resources: []*anypb.Any{{Value: []byte("test")}},
+			}
+
+			if tt.wantPanic {
+				defer func() {
+					if r := recover(); r == nil {
```
Comment (eshitachandwani) [inline internal/xds/clients/xdsclient/channel_test.go]: I am not sure if using recover in the test for the code path that also uses recover is a good way, but I dont have a better recommendation. 
Also we should check that the returned recover value contains the correct panic string to make sure we catch the correct panic. Or else the test could also pass when there is a panic because of the test setup.

File: internal/envconfig/envconfig.go [hunk 9]
```diff
@@ -96,6 +96,17 @@ var (
...
+	// catch unexpected panics.
+	//
+	// This feature can be disabled by setting the environment variable
+	// "GRPC_EXPERIMENTAL_XDS_PANIC_RECOVERY" to "false".
+	XDSRecoverPanicInResourceParsing = boolFromEnv("GRPC_EXPERIMENTAL_XDS_PANIC_RECOVERY", true)
```
Comment (eshitachandwani) [inline internal/envconfig/envconfig.go]: Can we change the env variable to specify that it only catches panics in resource parsing and not in all of xds.
Something like : GRPC_EXPERIMENTAL_XDS_RESOURCE_PANIC_RECOVERY"

File: internal/xds/clients/xdsclient/channel_test.go (hunk 6 above)
Comment (Pranjali-2501) [reply internal/xds/clients/xdsclient/channel_test.go]: The logic in else block is required for the case when env variable is set to true and decoderesponse will handle the panic occured.
We also need to assert the error returned while panic occurs and handled using recover().

File: internal/xds/clients/xdsclient/channel_test.go (hunk 8 above)
Comment (Pranjali-2501) [reply internal/xds/clients/xdsclient/channel_test.go]: >I am not sure if using recover in the test for the code path that also uses recover is a good way, but I dont have a better recommendation.

I could not think of any way other than recover to handle a panic scenerio in test. 

>Also we should check that the returned recover value contains the correct panic string to make sure we catch the correct panic. Or else the test could also pass when there is a panic because of the test setup.

Done

File: internal/envconfig/envconfig.go (hunk 9 above)
Comment (Pranjali-2501) [reply internal/envconfig/envconfig.go]: Done.

File: internal/xds/clients/xdsclient/channel_test.go (hunk 7 above)
Comment (Pranjali-2501) [reply internal/xds/clients/xdsclient/channel_test.go]: Done.

File: internal/xds/clients/xdsclient/channel_test.go (hunk 5 above)
Comment (Pranjali-2501) [reply internal/xds/clients/xdsclient/channel_test.go]: Done.

File: internal/xds/clients/xdsclient/channel_test.go (hunk 4 above)
Comment (Pranjali-2501) [reply internal/xds/clients/xdsclient/channel_test.go]: Done.

File: internal/envconfig/envconfig.go (hunk 3 above)
Comment (Pranjali-2501) [reply internal/envconfig/envconfig.go]: Modified the comment.

File: internal/xds/clients/xdsclient/channel_test.go (hunk 6 above)
Comment (eshitachandwani) [reply internal/xds/clients/xdsclient/channel_test.go]: What I meant was if we return after the `decodeResponse(opts, rType, resp)` in the `if tt.wantPanic ` block, we can write the else part of code without it being in an else block reducing the indentation.
basically do something like : 

```
			if tt.wantPanic {
				defer func() {
					if r := recover(); r == nil || !strings.Contains(fmt.Sprint(r), tt.wantErr) {
						t.Errorf("Expected panic want: %q, got: %q", tt.wantErr, r)
					}
				}()
				decodeResponse(opts, rType, resp)
				return
			} 
			if _, _, err := decodeResponse(opts, rType, resp); err == nil || !strings.Contains(err.Error(), tt.wantErr) {
			t.Fatalf("decodeResponse() failed with err: %v, want %q", err, tt.wantErr)
				
```

File: internal/xds/clients/xdsclient/channel_test.go [hunk 10]
```diff
@@ -772,3 +774,60 @@ func (ta *testEventHandler) waitForResourceDoesNotExist(ctx context.Context) (Re
...
+				Decoder:  panicDecoder{},
+			}
+			resp := response{
+				resources: []*anypb.Any{{Value: []byte("test")}},
+			}
+
+			if tt.wantPanic {
+				defer func() {
+					if r := recover(); r == nil || !strings.Contains(fmt.Sprint(r), tt.wantErr) {
+						t.Errorf("Expected panic want: %q, got: %q", tt.wantErr, r)
```
Comment (eshitachandwani) [inline internal/xds/clients/xdsclient/channel_test.go]: We generally prefer writing got before want.

Comment (Pranjali-2501) [reply internal/xds/clients/xdsclient/channel_test.go]: Done

File: internal/xds/clients/xdsclient/channel_test.go (hunk 6 above)
Comment (Pranjali-2501) [reply internal/xds/clients/xdsclient/channel_test.go]: Made the changes.

File: internal/envconfig/envconfig.go [hunk 11]
```diff
@@ -96,6 +96,14 @@ var (
 	// feature can be disabled by setting the environment variable
 	// GRPC_EXPERIMENTAL_PF_WEIGHTED_SHUFFLING to "false".
 	PickFirstWeightedShuffling = boolFromEnv("GRPC_EXPERIMENTAL_PF_WEIGHTED_SHUFFLING", true)
+
+	// XDSRecoverPanicInResourceParsing indicates whether the xDS resolver should
```
Comment (easwars) [inline internal/envconfig/envconfig.go]: This is not the resolver. This is the xds client resource parsing logic.

File: internal/xds/clients/xdsclient/channel.go [hunk 12]
```diff
@@ -253,7 +255,16 @@ func decodeResponse(opts *DecodeOptions, rType *ResourceType, resp response) (ma
 	perResourceErrors := make(map[string]error) // Tracks resource validation errors, where we have a resource name.
 	ret := make(map[string]dataAndErrTuple)     // Return result, a map from resource name to either resource data or error.
 	for _, r := range resp.resources {
-		result, err := rType.Decoder.Decode(NewAnyProto(r), *opts)
+		result, err := func() (res *DecodeResult, err error) {
+			defer func() {
+				if envconfig.XDSRecoverPanicInResourceParsing {
+					if p := recover(); p != nil {
+						err = fmt.Errorf("panic during %v resource decoding: %v\nstack: %s", rType.TypeName, p, debug.Stack())
```
Comment (easwars) [inline internal/xds/clients/xdsclient/channel.go]: Nit: This should say "recovered from a panic" instead of just saying "panic".

Also, consider including the complete resource in the error instead of simply the resource type name.

Does the stack contain anything useful? Is it the stack from where the panic happened or will it return the stack from the current goroutine, in which case it is not going to be useful.

File: internal/xds/clients/xdsclient/channel_test.go [hunk 13]
```diff
@@ -772,3 +774,60 @@ func (ta *testEventHandler) waitForResourceDoesNotExist(ctx context.Context) (Re
...
+
+func (panicDecoder) Decode(*AnyProto, DecodeOptions) (*DecodeResult, error) {
+	panic("simulate panic")
+}
+
+// TestDecodeResponse_PanicRecovery tests the panic recovery mechanism in
+// decodeResponse. It verifies that when XDSRecoverPanicInResourceParsing env
+// variable is enabled, panics during unmarshaling are caught and returned as
+// errors. When the env variable is disabled, it ensures the panic propagates.
+func (s) TestDecodeResponse_PanicRecovery(t *testing.T) {
```
Comment (easwars) [inline internal/xds/clients/xdsclient/channel_test.go]: Please consider splitting this into separate tests, since the main test logic is very different in each of these. See: https://g3doc.corp.google.com/company/teams/go-community/gotip/episodes/50.md?cl=head

File: internal/xds/clients/xdsclient/channel_test.go [hunk 14]
```diff
@@ -772,3 +774,60 @@ func (ta *testEventHandler) waitForResourceDoesNotExist(ctx context.Context) (Re
...
+	tests := []struct {
+		name          string
+		enableRecover bool
+		wantPanic     bool
+		wantErr       string
```
Comment (easwars) [inline internal/xds/clients/xdsclient/channel_test.go]: I don't think it is very valuable to check the exact error string being returned. Please correct me if you think otherwise.

File: internal/xds/clients/xdsclient/channel_test.go [hunk 15]
```diff
@@ -44,6 +45,7 @@ import (
 	v3routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
 	v3httppb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
 	v3discoverypb "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	utils "google.golang.org/grpc/internal/testutils"
```
Comment (easwars) [inline internal/xds/clients/xdsclient/channel_test.go]: You should generally try to rename the package that is closer to the current package. So, ideally we should be renaming `google.golang.org/grpc/internal/xds/clients/internal/testutils` to something like `xdstestutils` and let this one be `testutils`. 

See: https://docs.google.com/presentation/d/1qDeifaxW5AgWnxR6P6pT_gPu0zwes7dPdVQw6qGNN3Q/edit?slide=id.p#slide=id.p
go/go-style/best-practices#util-packages

Comment (Pranjali-2501) [reply internal/xds/clients/xdsclient/channel_test.go]: Done.

File: internal/envconfig/envconfig.go (hunk 11 above)
Comment (Pranjali-2501) [reply internal/envconfig/envconfig.go]: Done.

File: internal/xds/clients/xdsclient/channel.go (hunk 12 above)
Comment (Pranjali-2501) [reply internal/xds/clients/xdsclient/channel.go]: No, the stack will return the trace of current goroutine.

I have removed it and updated the error to include complete resource instead of resourceType name.

File: internal/xds/clients/xdsclient/channel_test.go (hunk 13 above)
Comment (Pranjali-2501) [reply internal/xds/clients/xdsclient/channel_test.go]: Done.

File: internal/xds/clients/xdsclient/channel_test.go (hunk 14 above)
Comment (Pranjali-2501) [reply internal/xds/clients/xdsclient/channel_test.go]: I think it is better to verify that panic occurred from the correct place. 

Also in case when the env variable is enabled and we want to verify that a panic occurs and caught by recover block. I think it is better to assert that the test is failing with correct error.

What do you think?
